static/app/utils/useRequest.tsx

```diff
@@ -0,0 +1,234 @@
+   * Teams that are not in the initial response.
+   */
+  hasMore: null | boolean;
+  /**
+   * The last query we searched. Used to validate the cursor
+   */
+  lastSearch: null | string;
+  /**
+   * Pagination
+   */
+  nextCursor?: null | string;
+};
+
+export type Result = {
+  /**
+   * This is an action provided to consumers for them to update the current
+   * result set using a simple search query.
+   *
+   * Will always add new options into the store.
+   */
+  onSearch: (searchTerm: string) => Promise<any>;
+  nextPage: () => Promise<void>;
+  get: () => Promise<any>;
+  post: (_) => Promise<any>;
+  update: (_) => Promise<any>;
+  del: () => Promise<any>;
+} & State;
+
+type Options = {
+  /**
+   * Number of items to return per page
+   */
+  limit?: number;
+};
+
+type FetchRequestOptions = {
+  limit?: Options['limit'];
+  cursor?: State['nextCursor'];
+  search?: State['lastSearch'];
+  lastSearch?: State['lastSearch'];
+};
+
+/**
+ * Helper function to actually load items
+ */
+async function fetch(
+  api: Client,
+  url: string,
+  {search, limit, lastSearch, cursor}: FetchRequestOptions = {}
+) {
+  const query: {
+    query?: string;
+    cursor?: typeof cursor;
+    per_page?: number;
+  } = {};
+
+  if (search) {
+    query.query = `${query.query ?? ''} ${search}`.trim();
+  }
+
+  const isSameSearch = lastSearch === search || (!lastSearch && !search);
+
+  if (isSameSearch && cursor) {
+    query.cursor = cursor;
+  }
+
+  if (limit !== undefined) {
+    query.per_page = limit;
+  }
+
+  let hasMore: null | boolean = false;
+  let nextCursor: null | string = null;
+  const [data, , resp] = await api.requestPromise(url, {
+    includeAllArgs: true,
+    query,
+  });
+
+  const pageLinks = resp?.getResponseHeader('Link');
+  if (pageLinks) {
+    const paginationObject = parseLinkHeader(pageLinks);
+    hasMore = paginationObject?.next?.results || paginationObject?.previous?.results;
+    nextCursor = paginationObject?.next?.cursor;
+  }
+
+  return {results: data, hasMore, nextCursor};
+}
+
+function useRequest(url: string, {limit}: Options = {}) {
+  const api = useApi();
+
+  const [state, setState] = useState<State>({
+    data: null,
+    loading: false,
+    hasMore: null,
+    lastSearch: null,
+    nextCursor: null,
+    error: null,
+  });
+
+  async function get(cursor: string | null = null) {
+    setState({...state, loading: true});
+    try {
+      const {results, hasMore, nextCursor} = await fetch(api, url, {
+        limit,
+        cursor,
+      });
+      setState({
+        ...state,
+        data: results,
+        hasMore,
+        loading: false,
+        nextCursor,
+      });
+      return Promise.resolve(results);
+    } catch (err) {
+      Sentry.captureException(err);
+      setState({...state, loading: false, error: err});
+      return Promise.reject(err);
+    }
+  }
+
+  async function handleSearch(search: string) {
```

Comment:
I think we should make a useDebounce hook and only call the search if the user has stopped typing. This will be a future PR.